### ISSUE TITLE
Bug in `tests/test_search.py/test_search_columns_with_iterables`

### ISSUE THREAD

**Issue by charles-turner-1** [ACCESS-NRI/intake-dataframe-catalog#68]
# Describe the bug
In the 3rd parametrization of `tests/test_search/test_search_columns_with_iterables.py`, we have the following:
```python
def test_search_columns_with_iterables(query, require_all, expected):
    df = pd.DataFrame(
        {
            "A": ["cat0", "cat1", "cat1"],
            "B": [["a", "b"], ["a", "b", "c"], ["c", "d", "a"]],
            "C": [("cx", "cy"), ("cx", "cz"), ("cz", "cy")],
            "D": [{0, 1}, {0, 2}, {0, 1}],
            "E": ["xxx", "xxx", "yyy"],
        }
    )
    results = search(
        df=df,
        query=query,
        columns_with_iterables=["B", "C", "D"],
        name_column="A",
        require_all=require_all,
    ).to_dict(orient="records")
    assert results == expected
```
where 
```python 
query =  {'B': ['a', 'b'], 'D': [0]}
require_all =  True
```
This means we start with the following dataframe:
```python
      A          B         C       D    E
0  cat0     [a, b]  (cx, cy)  {0, 1}  xxx
1  cat1  [a, b, c]  (cx, cz)  {0, 2}  xxx
2  cat1  [c, d, a]  (cz, cy)  {0, 1}  yyy
```
and apply the search parameters above to it. In the test, we have 
```python
>>> expected
      A       B         C    D    E
0  cat0  [a, b]  (cx, cy)  {0}  xxx
1  cat1  [a, b]  (cx, cz)  {0}  xxx
2  cat1     [a]  (cz, cy)  {0}  yyy
```
ie. all rows match - however `df.loc[2, 'B']` returns a list filtered for elements contained in the query. However, we have `require_all=True`, and so as `df.loc[2, 'B'] != query['B']`, this row should not be in the returned dataframe.

I believe that the test should expect this returned dataframe
```python
      A       B         C    D    E
0  cat0  [a, b]  (cx, cy)  {0}  xxx
1  cat1  [a, b]  (cx, cz)  {0}  xxx
```
# Additional context
I came across this bug whilst trying to wrap my head around this package. The `search._search` function is pretty difficult to understand & so I rewrote it from scratch in [polars](https://docs.pola.rs/api/python/stable/reference/index.html) as an exercise in understanding how it functions. I've been unable to reconcile this test failure with the other tests in my reimplementation.

**Comment by dougiesquire** [comment]
I think the test is correct. `name_column` is `"A"` and the value in the `"A"` column is `cat1` for both rows 1 and 2. So to include everything that satisfies the `require_all` criteria, I think row 2 needs to also be included.

I agree this function is confusing as heck and needs refactored. Would it be easiest to have a quick chat about requirements/functionality?

**Comment by charles-turner-1** [comment]
Yeah, I think that would be really helpful - I've been trying reverse engineer the logic & was struggling to get everything to match up.

With that said, I think I had misunderstood the intention of `require_all` - I've altered a couple of other tests where it also seemed inconsistent. I'm gonna have another crack first and see if that restores the tests to their previous state.

**Comment by charles-turner-1** [comment]
@dougiesquire I haven't been able to fix this - I can easily get the grouping to work as (I think) you've described above, but doing so breaks other tests, so I must be missing something important.

When would you be available to run me through the logic of how the search flags work? My calendar is generally pretty open for this week & next - I figure you've probably got a few pressing demands on your time so happy to fit around you.

 I'll be catching up with @rbeucher & @marc-white at 8AM AWST/ 11AM Hobart time tomorrow if you're available then?

**Comment by dougiesquire** [comment]
It was great to catch up with you @charles-turner-1! Just to try clarify (hopefully) and document a little what we discussed, I'll use your above example but ditch the `C` column for convenience.

So we have this query:

```python
query =  {'B': ['a', 'b'], 'D': [0]}
require_all =  True
```
on this dataframe:
```python
      A          B       D    E
0  cat0     [a, b]  {0, 1}  xxx
1  cat1  [a, b, c]  {0, 2}  xxx
2  cat1  [c, d, a]  {0, 1}  yyy
```



Both `cat0` and `cat1` satisfy the criteria since they contain both `B=a, D=0` and `B=b, D=0`. But `cat1` contains `B=a, D=0` with both `E=xxx` _and_ `E=yyy`. Both should be returned, so:

```python
      A       B    D    E
0  cat0  [a, b]  {0}  xxx
1  cat1  [a, b]  {0}  xxx
2  cat1     [a]  {0}  yyy
```

To try make this more practical, imagine `B` represents variable, `D` represents realm and `E` represents frequency. A user would want to know that `B=a, D=0` is available at multiple frequencies.

**Comment by charles-turner-1** [comment]
@dougiesquire thanks very much for the help! I've changed the implementation as you've suggested & it's resolved the questions with require_all.

I still can't quite understand what we expect to happen when we have`require_all=True` on non-iterable columns: see [CI run](https://github.com/ACCESS-NRI/intake-dataframe-catalog/actions/runs/11714952627/job/32630508082?pr=69).

I think in this instance, we would want to aggregate the column on name and then filter, rather than filter and then aggregate as we do on the columns? See also [comments on PR](https://github.com/ACCESS-NRI/intake-dataframe-catalog/pull/69#discussion_r1830384959)

**Comment by dougiesquire** [comment]
Could we just cast these to len 1 lists, agg and filter, then cast them back? (Note I'm on my phone atm and haven't looked at your links yet so might be missing something)

**Comment by charles-turner-1** [comment]
I did think about that - we've been doing that in the translators, so maybe that's been implicitly happening? It's a good starting point though.
